### PR TITLE
Inbox notification: Add undo snackbar button after deletion

### PR DESCRIPTION
**Pull request by octaedro**
Fixes #partially-4099

This PR adds the `undo` functionality after deletion.

_Please be sure the [PR 4262](https://github.com/woocommerce/woocommerce-admin/pull/4262) has been approved before approving this one_

> **_NOTE:_**  As this is a part of a big functionality this branch will be merged with the branch `add/4099`, and it will be that branch that will be merged with `master`.
To make the code reviewing easier, this PR only has the code referring to the functionality described in the title.
If you want to test everything together, all the changes are merged in the branch: `add/4099_test_branch`.
Run `unit test` directly from `add/4099_test_branch`.

### Screenshots
Single dismiss undo:
Screen Capture on 2020-05-04 at 14-17-35

Dismiss all undo:
Screen Capture on 2020-05-04 at 14-19-28

### Detailed test instructions:

1. To be able to see the inbox notifications, press Inbox in the navigation bar.
2. The list of inbox notifications should be visible.
2. Press `Dismiss` in any message.

**Flow A**
3. Choose `Dismiss this message`.
4. A confirmation modal will be visible. Press `Yes, I'm sure`.
5. A snackbar should be visible with the text: `Message dismissed.` and with an `Undo` button.
6. Press `Undo`. The message should be present in the list again.

**Flow B**
3. Choose `Dismiss all messages`.
- The behavior should be the same as choosing `Dismiss this message` but will affect all the messages.

### Changelog Note:
Enhancement: Added the `undo` functionality after deletion.

### CODE REVIEW

File: tests/api/admin-notes.php [hunk 1]
```diff
@@ -314,5 +316,100 @@ public function test_get_notes_schema() {
 		$this->assertArrayHasKey( 'date_reminder_gmt', $properties );
 		$this->assertArrayHasKey( 'actions', $properties );
 		$this->assertArrayHasKey( 'is_snoozable', $properties );
+		$this->assertArrayHasKey( 'is_deleted', $properties );
+	}
+
+	/**
```
Comment (timmyc) [inline tests/api/admin-notes.php:322]: Thanks for the tests <3

File: src/API/Notes.php [hunk 2]
```diff
@@ -227,6 +240,28 @@ public function update_item( $request ) {
...
+	 * Undo delete all notes.
+	 *
+	 * @param WP_REST_Request $request Request object.
+	 * @return WP_REST_Request|WP_Error
+	 */
+	public function undoremove_items( $request ) {
+		$data      = array();
+		$notes_ids = $request->get_param( 'notesIds' );
+		foreach ( (array) $notes_ids as $note_id ) {
+			$note = WC_Admin_Notes::get_note( $note_id );
```
Comment (timmyc) [inline src/API/Notes.php:253]: Should we cast $note_id to an (int) just to be extra safe?

File: client/wc-api/notes/mutations.js [hunk 3]
```diff
@@ -5,6 +5,74 @@ const updateNote = ( operations ) => ( noteId, noteFields ) => {
...
 };
 
+const removeNote = ( operations ) => async ( noteId ) => {
+	const { createNotice } = dispatch( 'core/notices' );
+	const resourceKey = 'note';
+	const resourceName = getResourceName( 'note', noteId );
+	const result = await operations.remove( [ resourceKey ], {
+		[ resourceKey ]: { noteId },
+	} );
+	const response = result[ 0 ][ resourceName ];
```
[the remark this answers is not in the document]
Comment (octaedro) [reply client/wc-api/notes/mutations.js]: Nice, it looks like a good idea to make it defensive. I had followed the pattern we used in other mutations, but I think that this is a better approach. I added it to the commit [719cbb9](https://github.com/woocommerce/woocommerce-admin/pull/4281/commits/719cbb963c4fcb8fd4e275e10dddc326dcfbb7ee).

File: src/API/Notes.php (hunk 2 above)
Comment (octaedro) [reply src/API/Notes.php]: Good idea! Sounds good to me. I added this to the commit [926424c](https://github.com/woocommerce/woocommerce-admin/pull/4281/commits/926424cb334572f79d87a89d88e358ffc21f185d).

**Comment by octaedro** [comment]
Hi @timmyc 

>It feels like we need some form of visual confirmation that an "undo" process is in-flight. Right now there is no visual indicator that something is happening until the notif appears again in the inbox. @jameskoster should we hide the original delete snackbar when "undo" is selected and display a new one? Might have a snackbar overload if we do that. Maybe we can change the "Undo" text to "restoring item..." or something.

Just as a PoC, I added a new `snackbar` with the text "Restoring item..." and did a gif to show you. It didn't look bad to me but I can make all the necessary changes.

Screen Capture on 2020-05-18 at 12-26-00


> I'm unable to test the undo on dismiss all, as on the test branch right now, I'm getting the following error when I do a dismiss all:

Yes, there was a bug in a fix I did. It was solved in the commit [b953a36](https://github.com/woocommerce/woocommerce-admin/pull/4262/commits/b953a367e53c5c4f6e189f77b5fab2da51d3626a). Now it should work normally.

File: tests/api/admin-notes.php [hunk 4]
```diff
@@ -314,5 +316,100 @@ public function test_get_notes_schema() {
...
+	public function test_undo_all_notes_delete() {
+		wp_set_current_user( $this->user );
+
+		$response = $this->server->dispatch( new WP_REST_Request( 'GET', $this->endpoint ) );
+
+		if ( $response->get_status() === 200 ) {
+			$response = $this->server->dispatch( new WP_REST_Request( 'DELETE', $this->endpoint . '/delete/all' ) );
+			$notes    = $response->get_data();
+			$this->assertEquals( 2, count( $notes ) );
+		} else {
```
Comment (becdetat) [inline tests/api/admin-notes.php:397]: How does this work, in a test? How do you get a 404?

**Comment by jameskoster** [comment]
> It feels like we need some form of visual confirmation that an "undo" process is in-flight.

I suppose the easiest and most consistent thing would be to display the notice in the loading state:

reload

File: tests/api/admin-notes.php (hunk 4 above)
Comment (octaedro) [reply tests/api/admin-notes.php:397]: You are completely right, that control is not necessary here.
I removed it in the commit [8b183e9](https://github.com/woocommerce/woocommerce-admin/pull/4386/commits/8b183e9e52698fdf831077fbc9e5a33bc231d3c5) in the [PR 4386](https://github.com/woocommerce/woocommerce-admin/pull/4386)
Thank you 🙂

**Comment by octaedro** [comment]
Hi @jameskoster,

> I suppose the easiest and most consistent thing would be to display the notice in the loading state:

If it is ok for you, I would like to address this design change in the [PR 4379](https://github.com/woocommerce/woocommerce-admin/pull/4379) to use the lastest card versions and wrap up this PR.

**Pull request merged**